### PR TITLE
JobRequirement refactor

### PR DESCRIPTION
**Pull request by TheShuEd**
## About the PR
Role requirements have been greatly improved. Adding new requirements is now much more convenient, obsolete syntax has been removed, and support for character profile restrictions has been added. also now any rules can be inverted, not just some.

As an example, I added a character age restriction that characters under 20 can't be plugged into command roles. This will also allow forks to use the system to restrict roles by race, gender, and other profile parameters.

also added species restriction, but not used ingame

## Why / Balance
I saw the shitcode, I fixed the shitode.
also wanna use species requirement in my fork

## Media

https://github.com/user-attachments/assets/40b4995e-9a72-40ee-ae0d-45412e4a484f

bonus: also work with loadouts (just example, not added into game)
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- tweak: Nanotrasen has introduced recruitment rules. Characters under the age of 20 are no longer allowed to take on the role of head.

### CODE REVIEW

**Comment by lzk228** [comment]
remove rp ahh changelog

**Comment by kel593** [comment]
veri want to see a species jobblock

**Comment by chavonadelal** [comment]
It's cool, but oh, and they forgot about localization again... I hope my pull request with localization of the same system will be accepted in #30208   ..

**Comment by chavonadelal** [comment]
Also #18425

**Comment by TheShuEd** [comment]
I didn't change the logic itself within the existing constraints, just moved it to a new format

File: Content.Shared/Roles/DepartmentPrototype.cs [hunk 1]
```diff
@@ -9,7 +9,13 @@ public sealed partial class DepartmentPrototype : IPrototype
...
+    /// The name LocId of the department that will be displayed in the various menus.
+    /// </summary>
+    [DataField(required: true)]
+    public LocId Name = string.Empty;
+
+    /// <summary>
+    /// A description LocId to display in the character menu as an explanation of the department's function.
     /// </summary>
     [DataField(required: true)]
     public string Description = string.Empty;
```
Comment (chavonadelal) [inline Content.Shared/Roles/DepartmentPrototype.cs:21]: The summary says that the Description is a LocId but it is still just a string

Comment (TheShuEd) [reply Content.Shared/Roles/DepartmentPrototype.cs:21]: oh noooo

Comment (TheShuEd) [reply Content.Shared/Roles/DepartmentPrototype.cs]: fixed

**Comment by qrwas** [comment]
I have a question about the age limit. Does it just check if the age is equal to the specified one? have you considered adding ranges? 
And the same for the race restriction, if you plan to add them, will you consider adding multiple races to the restriction or allowed?

**Comment by TheShuEd** [comment]
> I have a question about the age limit. Does it just check if the age is equal to the specified one? have you considered adding ranges?

for age, there is a check to see if the age is greater or less than a certain value. If you need a specific range of ages, you can use 2 such restrictions per role, with different settings. 

> And the same for the race restriction, if you plan to add them, will you consider adding multiple races to the restriction or allowed?

I don't plan on adding them yet, considering wizden has already rejected it. Specific implementations will already depend on forks, but it really would be very easy to do it as a list of allowed or blocked species.

**Comment by TheShuEd** [comment]
ok, I talked toEemo, and I'll add species restrictions here, even though they won't be used in the game.

**Comment by TheShuEd** [comment]
image
image

**Comment by qrwas** [comment]
> for age, there is a check to see if the age is greater or less than a certain value. If you need a specific range of ages, you can use 2 such restrictions per role, with different settings.

Can you please show us an example of how this works?
Wouldn't it be easier to add 2 variables: minimum age, maximum age?

**Comment by TheShuEd** [comment]
```yml
requirements:
  - !type:AgeRequirement #minimum 20
    requiredAge: 20
  - !type:AgeRequirement #maximum 40
    requiredAge: 20
    inverted: true
    ```

**Comment by qrwas** [comment]
> ```yaml
> requirements:
>   - !type:AgeRequirement #minimum 20
>     requiredAge: 20
>   - !type:AgeRequirement #maximum 40
>     requiredAge: 40
>     inverted: true
> ```
Cool. Thanks 👍

File: Content.Shared/Roles/JobRequirement/SpeciesRequirement.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        HumanoidCharacterProfile? profile,
+        IReadOnlyDictionary<string, TimeSpan> playTimes,
+        [NotNullWhen(false)] out FormattedMessage? reason)
+    {
+        reason = new FormattedMessage();
+
+        if (profile is null) //the profile could be null if the player is a ghost. In this case we don't need to block the role selection for ghostrole
+            return true;
+
+        var speciesList = "[color=yellow]";
```
Comment (metalgearsloth) [inline Content.Shared/Roles/JobRequirement/SpeciesRequirement.cs:32]: Use stringbuilder because these allocate every time.

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**